File: frontend/src/schema/commands.ts
/**
 * The Canvas Command protocol — the single mutation API for the canvas.
 *
 * Toolbar clicks, the Canvas Commander agent, and the P&ID Extractor all
 * produce these same commands. The frontend store and the backend command bus
 * each implement an `apply(state, command)` reducer over this union, so AI and
 * humans share one authoritative state model.
 */

import type { CanvasGroup, CanvasNode, CanvasState, PipeData, Position } from './canvas';
import type { EquipmentType } from './equipment';

export interface AddNodeCommand {
  op: 'add_node';
  id: string;
  equipment: EquipmentType;
  position: Position;
  label?: string;
  rotation?: number;
  data?: Record<string, unknown>;
}

export interface UpdateNodeCommand {
  op: 'update_node';
  id: string;
  patch: Partial<Pick<CanvasNode, 'label' | 'rotation' | 'data'>>;
}

export interface MoveNodeCommand {
  op: 'move_node';
  id: string;
  position: Position;
}

export interface RemoveNodeCommand {
  op: 'remove_node';
  id: string;
}

export interface ConnectCommand {
  op: 'connect';
  id: string;
  source: string;
  sourcePort: string;
  target: string;
  targetPort: string;
  data?: PipeData;
}

export interface UpdateEdgeCommand {
  op: 'update_edge';
  id: string;
  patch: Partial<PipeData>;
}

export interface DisconnectCommand {
  op: 'disconnect';
  id: string;
}

export interface SelectCommand {
  op: 'select';
  ids: string[];
}

export interface ClearCommand {
  op: 'clear';
}

export interface AddGroupCommand {
  op: 'add_group';
  id: string;
  label: string;
  position: Position;
  width: number;
  height: number;
  memberIds: string[];
  color?: string;
}

export interface UpdateGroupCommand {
  op: 'update_group';
  id: string;
  patch: Partial<Pick<CanvasGroup, 'label' | 'color' | 'position' | 'width' | 'height' | 'memberIds'>>;
}

export interface RemoveGroupCommand {
  op: 'remove_group';
  id: string;
}

/** Atomic group — applied all-or-nothing. Used by the P&ID Extractor. */
export interface BatchCommand {
  op: 'batch';
  commands: CanvasCommand[];
}

export type CanvasCommand =
  | AddNodeCommand
  | UpdateNodeCommand
  | MoveNodeCommand
  | RemoveNodeCommand
  | ConnectCommand
  | UpdateEdgeCommand
  | DisconnectCommand
  | SelectCommand
  | ClearCommand
  | AddGroupCommand
  | UpdateGroupCommand
  | RemoveGroupCommand
  | BatchCommand;

export type CommandOp = CanvasCommand['op'];

/** Who issued a command — used for UI attribution and logging. */
export type CommandSource =
  'user' | 'agent:optimus' | 'agent:pid_extractor' | 'agent:commander' | 'system';

/** WebSocket message envelopes (server <-> client). */
export interface CommandMessage {
  type: 'command';
  source: CommandSource;
  command: CanvasCommand;
}

export interface SnapshotMessage {
  type: 'snapshot';
  state: CanvasState;
}

export type ServerMessage = CommandMessage | SnapshotMessage;
export type ClientMessage = CommandMessage;

File: backend/src/inf_canvas/schema/commands.py
"""Canvas Command protocol (Pydantic mirror of commands.ts).

A discriminated union on `op`; `command_adapter` validates raw dicts coming off
the wire into the correct member type.
"""

from __future__ import annotations

from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field, TypeAdapter

from .canvas import CanvasState, PipeData, Position
from .equipment import EquipmentType


class AddNodeCommand(BaseModel):
    op: Literal["add_node"]
    id: str
    equipment: EquipmentType
    position: Position
    label: str | None = None
    rotation: float | None = None
    data: dict[str, Any] | None = None


class UpdateNodePatch(BaseModel):
    label: str | None = None
    rotation: float | None = None
    data: dict[str, Any] | None = None


class UpdateNodeCommand(BaseModel):
    op: Literal["update_node"]
    id: str
    patch: UpdateNodePatch


class MoveNodeCommand(BaseModel):
    op: Literal["move_node"]
    id: str
    position: Position


class RemoveNodeCommand(BaseModel):
    op: Literal["remove_node"]
    id: str


class ConnectCommand(BaseModel):
    op: Literal["connect"]
    id: str
    source: str
    sourcePort: str
    target: str
    targetPort: str
    data: PipeData | None = None


class UpdateEdgeCommand(BaseModel):
    op: Literal["update_edge"]
    id: str
    patch: PipeData


class DisconnectCommand(BaseModel):
    op: Literal["disconnect"]
    id: str


class SelectCommand(BaseModel):
    op: Literal["select"]
    ids: list[str]


class ClearCommand(BaseModel):
    op: Literal["clear"]


class BatchCommand(BaseModel):
    op: Literal["batch"]
    commands: list[CanvasCommand]


CanvasCommand = Annotated[
    AddNodeCommand
    | UpdateNodeCommand
    | MoveNodeCommand
    | RemoveNodeCommand
    | ConnectCommand
    | UpdateEdgeCommand
    | DisconnectCommand
    | SelectCommand
    | ClearCommand
    | BatchCommand,
    Field(discriminator="op"),
]

BatchCommand.model_rebuild()

command_adapter: TypeAdapter[CanvasCommand] = TypeAdapter(CanvasCommand)

CommandSource = Literal[
    "user",
    "agent:optimus",
    "agent:pid_extractor",
    "agent:commander",
    "system",
]


class CommandMessage(BaseModel):
    type: Literal["command"] = "command"
    source: CommandSource
    command: CanvasCommand


class SnapshotMessage(BaseModel):
    type: Literal["snapshot"] = "snapshot"
    state: CanvasState


ServerMessage = CommandMessage | SnapshotMessage
